Convolutional model for seismic data (using a depth model)

victortxa opened the pull request:
Generate a convolutional model from a depth geological model (Vp; density optional). It's useful to put direct in geology instead of a model in time, avoiding lose the main information. It can be helpful in situations when it's boring to convert from depth to time.

## Checklist:

- [x] Make tests for new code
- [x] Create/update docstrings
- [x] Include relevant equations and citations in docstrings
- [x] Code follows PEP8 style conventions
- [x] Code and docs have been spellchecked
- [x] Include new dependencies in docs, requirements.txt, README, and .travis.yml
- [x] Documentation builds properly
- [x] All tests pass
- [ ] Can be merged
- [x] Changelog entry (leave for last)
- [x] First-time contributor? Add yourself to `doc/contributors.rst` (leave for last)

eusoubrasileiro commented:
Hi @victortxa. I am really curious about what does it really do. Does it create a synthetic seismic ? 
I know that documentation will be provided soon, just wanted to know the general ideia because I might have done something similar in the past.

Cheers and wellcome

victortxa commented:
Hi @eusoubrasileiro thank you. 
The goal is to take a Vp (density is optional) model in depth and generates the convolutional seismic response in time. In general, the Vp should be given in time (like here: http://www.agilegeoscience.com/blog/2013/12/12/to-make-a-wedge.html ), what makes 'harder' to think geologically.  The source is a ricker. @leouieda is introducing me to Git, so I make this pull request to begin, but I'll detail the function asap.
Cheers.

eusoubrasileiro commented:
@victortxa nice! some time ago I did [this](https://github.com/fatiando/fatiando/pull/61) I didn't know about Evan Bianco and agile geoscience. I did as a cookbook for testing obspy... Well but It was too much for just testing obpsy io in a cookbook.  

Now It will be really nice to have such a tool, really cool! 
Again good initiative  keep the good job!

Btw I guess that with so many seismic modules we are heading to need to restruture the seismic package to new inner packages? What do you think @leouieda?

leouieda commented:
@victortxa any progress on this?
 
Could you write an introductory paragraph at the start of the PR description? It should contain a summary of the changes you're making (maybe an example?). Code from #192 will be useful for this (should be merged soon).

victortxa commented:
@leouieda Should I write a description of the code or only about the changes I made? About changes do you mean what the code adds to Fatiando? #192 is going to be merged to Fatiando, right? Sorry, I'm still a little confused here.

leouieda commented:
Just a general description of what this PR adds. Nothing too big (2-3 sentences is fine).

#192 is practically done so it won't take too long to be merged. Once it is, you can updated your branch and use the plotting functions proposed there.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+"""
+
```
@victortxa the first line of the module docstring should describe in a single sentence what this module does. In this case, a good examples would be "Seismic forward modeling by convolution" or "Zero-offset convolutional seismic modeling".

Next, put a small paragraph describing what the functions here do. Very general.

Finally, put a list of the functions present here with a line saying what each does (usually this is the first line of the function docstring).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+"""
+
+"""
+import numpy as np
```
First thing you should import is `from __future__ import division`. This will enable the new division handling of Python 3. In Python 2 `1/2 == 0`. In Python 3 `1/2 == 0.5` and `1//2 == 0`. So `/` denotes division and `//` denotes integer division. This makes things better because you don't have to call `float` on everything.

victortxa replied:
Should I use Python 3 already?

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    for j in range(0, n_traces):
+        kk = np.ceil(TWT[0, j]/dt_dwn)
+        lim = np.ceil(TWT[-1, j]/dt_dwn)-1
+    # linear interpolation
```
This is nitpicking but you should avoid this kind of comment. I can see from the line below that you are doing linear interpolation. So the comment doesn't contribute much. Instead, use the comment to explain *why* you are doing a linear interpolation.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    # extension of the model repeats the last value of the true model
+        vel[lim:, j] = vel[lim-1, j]
+    # first values equal to the first value of the true model
+        vel[0:kk, j] = model[0, j]
```
Same thing for these two comments above as well.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    # first values equal to the first value of the true model
+        vel[0:kk, j] = model[0, j]
+    # resampling from dt_dwn to dt
+    vel_l = np.zeros((np.ceil(TMAX/dt), n_traces))
```
This one is great!

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    # resampling from dt_dwn to dt
+    vel_l = np.zeros((np.ceil(TMAX/dt), n_traces))
+    TWT_ts = np.zeros((np.ceil(TMAX/dt), n_traces))
+    resmpl = int(dt/dt_dwn)
```
As per the comment in the beginning of the file, if you want the integer division of `dt` and `dt_dwn`, use `dt//dt_dwn`. Use `int` if this needs to be an integer type (to use for indexing, for example).

victortxa replied:
I use this variable four lines below as array indexing, so I should keep the calculation like this, right?

leouieda replied:
OK, I hadn't seen it there. You can keep it like that or use integer division to make it more explicit.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+            for jj in range(1, len(TWT_ts)):
+                rho_l[jj, j] = rho[resmpl*jj, j]
+    # calculate RC
+    if ~isinstance(rho, np.ndarray):
```
Is this really a `~`? Usually `~` negates numpy arrays. I've never seen it used like this before. Could you use `not isinstance(...` instead?

victortxa replied:
I think I looked how to negate something in Python and found this. I thought it was valid for any variable, and it fit my expectations (I prefer a symbol than writing not).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+        rc[1:, :] = (vel_l[1:, :]-vel_l[:-1, :])/(vel_l[1:, :]+vel_l[:-1, :])
+    else:
+        rc[1:, :] = ((vel_l[1:, :]*rho_l[1:, :]-vel_l[:-1, :]*rho_l[:-1, :]) /
+                     (vel_l[1:, :]*rho_l[1:, :]+vel_l[:-1, :]*rho_l[:-1, :]))
```
So all of the above is time model building, right? It might be better to separate it in a different function then, like you did with the wavelet code. This will make this function easier to test. Specially since most of the code is model building, hence that's where it's most likely to have bugs. Fortunately, model building is easier to test that the actual results of convolution.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+        else:
+            aux = np.floor(len(w)/2.)
+            synth_l[:, j] = np.convolve(rc[:, j], w, mode='full')[aux:-aux]
+    return synth_l, TWT_ts
```
Could we separate the convolution part in another function as well? Imagine someone wants to use a time domain model. Then, this function would simply chain the call to depth->time conversion and time->convolution.

leouieda replied:
This would be good to have so people can interface with existing code to model stuff (you mentioned most people model in the time domain).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    return synth_l, TWT_ts
+
+
+def rickerwave(f, dt):
```
In Python, there is no concept of "private" functions (things that can only access from within a module/class). But there is a convention to let people know if a certain function/class should not be used in other places. For "private" functions, we append a `_` to the beginning of the function name, e.g. `_rickerwave`. This marks that this function is for internal use only. Such functions don't need to have very detailed docstrings, just a single line is enough.

So consider if this will be a private function for this module. If not, than you should make a complete docstring for it. It is probably better to have it be public so that people can plot the wavelet if they want.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,126 @@
+    * model : 2D-array
+        Vp values
+    * f : float
+        Dominant frequency of the ricker wavelet
```
Could you take the wavelet as input? Do people actually use other wavelets in seismics?

victortxa replied:
True! An excellent idea. I think I oversimplified this part.

leouieda commented:
@victortxa finally got a chance to review the code. Good job, it looks great!

leouieda commented:
@eusoubrasileiro if you have any free time, could you take a look at the code and see if you have any ideas/comments? You've done this kind of modeling before so you're the boss here this time :facepunch:

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,161 @@
+
+----
+
+"""
```
TravisCI [is complaining](https://travis-ci.org/fatiando/fatiando/builds/62279201#L670) of a non-ascii character. Probably the "Öz". To get Unicode characters, try putting this comment on the first line of the file:

    # -*- coding: utf-8 -*-

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,181 @@
+**References**
+
+Yilmaz, Oz,
+Ch.2 Deconvolution. In: YILMAZ, Öz. Seismic Data Analysis: Processing, 
```
You missed the Öz here as well.

victortxa replied:
Sorry!

leouieda commented:
@victortxa Great job! Leave a comment here when you're done pushing changes and I can review the code.

eusoubrasileiro commented:
@leouieda sorry I havent look this in a while, but I will try this weekend. Good Job by @victortxa

eusoubrasileiro commented:
@victortxa now I am following the thread. I can say for now, that would be nice to have the notebook you make your tests :-D and that comes to the cookbook example @leouieda any changes on this or do we still need a cookbook for every new module?
if that is the case a simple example of coobook is [here](https://github.com/fatiando/fatiando/blob/master/cookbook/vis_mpl_irregular_data.py) or others [here](https://github.com/fatiando/fatiando/blob/master/cookbook/seismic_wavefd_love_wave.py) i think fits better your case.

leouieda commented:
@eusoubrasileiro that velocity model (I think you call it pinch out) would be great for #202 ! Thanks for taking a look at this.

victortxa commented:
@eusoubrasileiro I am taking a look in your tips. I didn't put the 'notebook' I'm using because I use another function I haven't pushed (in true, a modified a version of AgileGeo, Evan Bianco, make_wedge function) and I want to include it probably in a new module for 'model building'. My goal is to put a simpler model for the test, maybe a two parallel layers.
What do you think?

victortxa commented:
@eusoubrasileiro I fixed some details now. I am gonna put the cookbook asap, I told you one done.

victortxa commented:
@eusoubrasileiro Example added in cookbook. Doing the example I noted some small fixes I'm gonna do. Take a look when you can.

victortxa commented:
![output_cookbook](https://cloud.githubusercontent.com/assets/10142902/8107550/807e3a78-1020-11e5-831f-5e8f7bed50f9.png)
Output of the cookbook. I think it can be improved.

eusoubrasileiro commented:
@victortxa  the example seams good! The picture is nice but it seams that the velocity in the last plot is the same all over the model? Is that right? I guess that just a two plot woud be sufficient. I will try find some time to look at the code. But thanks for doing this for fatiando we need more seismic!! hehe

eusoubrasileiro commented:
I will try help you soon, I think making the wavelet more visible would be nice, try changing the scale factor. But all the way around it's really nice... I wanna play with your code soon

victortxa commented:
There's a little mistake in one parameter of this plot. But, when I overlayer the wiggles with the correct parameter the same problem happens. Trying to find the mistake.

victortxa commented:
@eusoubrasileiro I fixed the picture generation, I changed the plot function from mpl.imshow to mpl.seismic_image, since I over print the wiggles. Your plot functions helped me to note that a time vector output from the function seismic_conv.depth_2_time wasn't necessary, since we have the dt.

![seismic_conv](https://cloud.githubusercontent.com/assets/10142902/8383933/b2506022-1c13-11e5-8be5-4a4d694bc452.png)

leouieda commented:
Would it be better to use less traces? Is there a reason for so many traces
(honest question)?
On Jun 26, 2015 2:57 PM, "victortxa" <notifications@github.com> wrote:

> @eusoubrasileiro <https://github.com/eusoubrasileiro> I fixed the picture
> generation, I changed the plot function from mpl.imshow to
> mpl.seismic_image, since I over print the wiggles. Your plot functions
> helped me to note that a time vector output from the function
> seismic_conv.depth_2_time wasn't necessary, since we have the dt.
>
> [image: seismic_conv]
> <https://cloud.githubusercontent.com/assets/10142902/8383933/b2506022-1c13-11e5-8be5-4a4d694bc452.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/190#issuecomment-115815223>.
>

eusoubrasileiro commented:
@victortxa nice I liked you fix It. 
I am in a lot o throuble here with brazil 13th round and my wedding but as soon as I get some time I'll take a look. 
@leouieda yes I dont think there is a need for that many also. Could be just some 20 traces I guess.

victortxa commented:
@leouieda Thank you, less traces are much better!
@eusoubrasileiro Don't worry, now I think only the tests must be done to finish. Concentrate in the 13th bid, everyone is waiting it!
![seismic_conv2](https://cloud.githubusercontent.com/assets/10142902/8455389/a26eba4e-1fda-11e5-816b-530c7954a002.png)

leouieda commented:
That looks great!
True, @eusoubrasileiro, focus on the important things :+1:

Victor, now you should focus on making some unit tests. Let me know if you
need help with that. This is a difficult topic but one of the most
important ones in sustainable software development.

On Wed, Jul 1, 2015 at 10:21 AM, victortxa <notifications@github.com> wrote:

> @leouieda <https://github.com/leouieda> Thank you, less traces are much
> better!
> @eusoubrasileiro <https://github.com/eusoubrasileiro> Don't worry, now I
> think only the tests must be done to finish. Concentrate in the 13th bid,
> everyone is waiting it!
> [image: seismic_conv2]
> <https://cloud.githubusercontent.com/assets/10142902/8455389/a26eba4e-1fda-11e5-816b-530c7954a002.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/190#issuecomment-117671568>.
>

leouieda commented:
@victortxa Travis CI was unable to push the docs to `fatiando/pull`. This seems to be a [problem with the git credentials](http://stackoverflow.com/questions/18027115/committing-via-travis-ci-failing). You'll need to make some changes to `ci-tools/update-website.sh`.

On lines 17 and 48, replace the git URLs `https://${GH_TOKEN}@github.com/...` with `https://leouieda:${GH_TOKEN}@github.com/...`.

This should work now. Try making these changes and pushing, then check if http://www.fatiando.org/pull/190 is working.

victortxa commented:
@leouieda It seems it didn't work. I saw that in #127 this file is a little different (without the changes in lines 17 and 48) and it's working. Any idea?

leouieda commented:
I think it works for the other Pull Requests because I made them. And Travis is using a github authentication token form my account, I think. I'm not finding much about this on Google.

For now, can you revert the last commit (nothing fancy, just make another commit to undo the changes)?

I'll see if I can find a fix.

eusoubrasileiro commented:
very cool

victortxa commented:
@leouieda Almost done. All functions are tested now. Although we can't trust in 'make coverage', I liked to see 100% there. :+1: 
I don't if it would be better to merge the test_rc_shorter_than_wavelet() in the test_impulse_response():.
Now I'll add the example in the main function header.

leouieda commented:
@victortxa the tests are looking great! Nice to see 100% coverage! (Feeling ashamed that I don't strive for that with my own code)

You can leave the tests separated. It's better to have them as "unit"-like as possible.

Is the example all that is missing? You can check a few items from the PR checklist.

#232 should fix the doctest error you're having.

victortxa commented:
@leouieda I can't remember what code you showed me with that good example that works as a test. I'd follow it.

leouieda commented:
@victortxa [it was this one](http://www.fatiando.org/api/gravmag.eqlayer.html#fatiando.gravmag.eqlayer.EQLGravity)

victortxa commented:
@leouieda In ```make view-docs``` only the cookbook is been shown, without the last modification I did. When I give ```make docs``` an error appears:
_Exception occurred:
_File "conf.py", line 12, in <module>
from fatiando import __version__, __commit__
_ImportError: cannot import name __commit__
Also in the seismic features page the ```conv```  function doesn't appear, although I've modified the file 
 ```/seismic/__init__ .py ```
Any tip?

leouieda commented:
@victortxa that's because sphinx needs to `import fatiando` to get the docstrings. So it'll build the docs using the version that it imports. You need to make sure that it is the version in your repository, not an installed version. I usually do this by setting `PYTHONPATH` but you can do it also with [conda environments](http://conda.pydata.org/docs/using/envs.html).

victortxa commented:
@leouieda  :+1: Thank you!  
Now make worked (```Converting seismic_conv.py to rst ... done```), but the cookbook is not being updated (during make ```make -rf cookbook``` is done...).

victortxa commented:
@leouieda I think it's all ok. There's only a small detail in docs (the input parameter in the rickerwave isn't appearing bold in the generated doc).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,245 @@
+    Parameters:
+
+    * f : dominant frequency value in Hz
+    * dt : time sampling rate in seconds (usually it is in the order of ms)
```
@victortxa please indicate the expected type of the input parameters (I'm assuming both are floats).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,245 @@
+    .. math::
+
+        f < 0.2 \frac{1}{2 dt}.
+
```
Could you add a bit more  about what the function actually does? From the code, I see that it returns an array with the wavelet at some calculated time  values. The user seems to have no control over what times the wavelet is calculated, right? So it would help  if you explain why and what times are being used  for calculation.

victortxa commented:
@leouieda I think/hope it's everything fine, finally...

leouieda commented:
@victortxa I'll have a look at this as soon as I can. Seems that all tests pass and there are no merge conflicts :+1: I just need to take a quick look at the code and see if there is anything missing that I can spot (probably not).

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+from scipy import interpolate  # linear interpolation of velocity/density
+
+
+def convolutional_model(rc, f, wavelet, dt=2.e-3):
```
@victortxa now that I'm thinking about this better, I don't think `dt` should have a default value. This is the sampling interval of `rc` and the results would be totally wrong if this value is wrong. So it's better to require the user to think about.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+    return synth_l
+
+
+def reflectivity(model_t, rho=1.):
```
As with `dt`, I don't think `rho`  should have a default value. Defaults are better when the user doesn't really  need to think about this, only for advanced use. This is not the case with `rho`. Forgetting to pass `rho` would lead  to wrong reflectivity.

victortxa replied:
@leouieda I put a default value in `rho` because in general only velocity is provided by the user, giving `rho` is more optional. Do you think it's better to put it as a needed input and just warn that put `1.0` to disconsider this parameter?

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+                     (model_t[1:, :]*rho[1:, :]+model_t[:-1, :]*rho[:-1, :]))
+    except TypeError:
+        rc[1:, :] = ((model_t[1:, :]-model_t[:-1, :]) /
+                     (model_t[1:, :]+model_t[:-1, :]))
```
Since `rho`  shouldn't have a default, you don't need the `try ...  except` blocks anymore. What you  should have is some `asserts` in the beginning to check if the shapes of `model_t` and `rho` match.

You should also check if both have 2 dimensions (you're assuming this in the calculations). Something like `assert rho.ndim == 2`.

victortxa replied:
@leouieda Rewriting the code I noticed one thing. Maybe the user is interested in calculate the reflectivity just for one trace. 
So, how do you think is the best way to keep this option (if you think it a good idea)? I realized to keep the `try`block, but in the `except` I would write considering `model_t` and `rho` as one dimensional. In this way,  `assert model_t.shape == rho.shape`  would replace `assert rho.ndim == 2`.

leouieda replied:
@victortxa since you're only indexing in the first dimension, you can omit the `:` for the column index:

    rc[1:] = ((model_t[1:]*rho[1:]-model_t[:-1]*rho[:-1]) /
                 (model_t[1:]*rho[1:]+model_t[:-1]*rho[:-1]))

This should work the same in 2D or 1D. No need to checking the dimensions then.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+def reflectivity(model_t, rho=1.):
+    """
+    Calculate reflectivity series
+
```
Could you add a few sentences here to say that this function expects things in the time-domain and the user should use `depth_2_time` to convert?

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+    Parameters:
+
+    * model : 2D-array
+        Vp values
```
Please add something like "in the depth domain" to make it clear.

leouieda commented on fatiando/seismic/conv.py:
```diff
@@ -0,0 +1,251 @@
+    return rc
+
+
+def depth_2_time(model, dt=2.e-3, dz=1., rho=1.):
```
Again, `dt` and `dz` should not have defaults because these will vary depending on the model.

Also, model doesn't have to be Vp. It could be any physical property given in depth that you want to convert to time. So instead of passing `model` as Vp and `rho` as well, the user can call this function twice to get vp and rho in time:

    vp_t = depth_2_time(vp, dt, dz)
    rho_t = depth_2_time(rho, dt, dz)

victortxa replied:
I agree, but I only did like this because I kept `rho` as an optional parameter. If we modify this (above), this adjust here will also be necessary. The negative side I see will be the necessity to call the `depth_2_time` twice, but if the user doesn't want (or have) any `rho` then it will not required. Maybe we could keep the default value in the `reflectivity`function and do this modification here in the `depth_2_time`. 
What do you think?

leouieda replied:
@victortxa I still prefer a more explicit interface. Having rho as default feels strange because it's a crucial parameter to calculate the impedance. It's usually better to be explicit so that users are required to think about that they are doing. In this case, that would mean calling `depth_2_time` twice or creating a `np.ones` array for `reflectivity`. Both are not too complicated.

leouieda commented:
@victortxa sorry I took so long to look at this. I made a few comments about the function parameters. Not too much to do, just a few tweaks to make the code and the API cleaner. You'll need to adapt the cookbook and tests to the required changes.

After this is done, I think it should be ready to merge. You'll need to merge the fatiando/fatiando master branch into this one before we can merge the PR.

victortxa commented:
@leouieda Now I took so long to answer, sorry. Maybe it is easier if I show you the modifications I mentioned above, it will be easier to understand. :grimacing:

victortxa commented:
@leouieda Now I guess it's all done.

leouieda commented:
@victortxa all seems good to go! Sorry for the long wait. Merging!

eusoubrasileiro commented:
Cool!

On Tue, Feb 2, 2016 at 3:02 PM, Leonardo Uieda <notifications@github.com>
wrote:

> Merged #190 <https://github.com/fatiando/fatiando/pull/190>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fatiando/fatiando/pull/190#event-536149905>.
>



-- 
Um   homem que fosse só homem e dissesse as coisas que Jesus disse não
seria um grande mestre de moral. Seria um lunático no mesmo nível de um
homem que diz ser um ovo cozido ou então seria o próprio diabo. Cada um de
nós precisa tomar a sua decisão. Ou este homem era, e é, o Filho de Deus,
ou então um louco, ou algo pior... Mas não venhamos com nenhum argumento
complacente que diga que ele foi apenas um grande mestre humano. Ele não
nos deu esta escolha. Nunca pretendeu fazê-lo. C. S. Lewis - Cristianismo
Puro e Simples

victortxa commented:
:+1: Fine! :smile:
